docs: DevExpressApp - Single project updates (#2296)
unoplatform/uno.extensions: doc/Learn/Maui/ThirdParty-DevExpress.md

## Getting Started

### [Visual Studio](#tab/vs)
 
> [!NOTE]
> If you don't have the **Uno Platform Extension for Visual Studio** installed, follow [these instructions](xref:Uno.GetStarted.vs2022).
 
- Launch **Visual Studio** and click on **Create new project** on the Start Window. Alternatively, if you're already in Visual Studio, click **New, Project** from the **File** menu.
 
- Type `Uno Platform` in the search box
 
- Click **Uno Platform App**, then **Next**
 
- Name the project `DevExpressApp` and click **Create**
 
At this point you'll enter the **Uno Platform Template Wizard**, giving you options to customize the generated application.
 
- Select **Blank** in **Presets** selection
 

> - Select the **Platforms** tab and unselect **MacCatalyst** and **Windows** platforms

Reviewer: Does WASM and Desktop support MauiEmbedding?

> - Launch **Visual Studio** and click on **Create new project** on the Start Window. Alternatively, if you're already in Visual Studio, click **New, Project** from the **File** menu.
>  
> - Type `Uno Platform` in the search box
>  
> - Click **Uno Platform App**, then **Next**
>  
> - Name the project `DevExpressApp` and click **Create**
>  
> At this point you'll enter the **Uno Platform Template Wizard**, giving you options to customize the generated application.
>  
> - Select **Blank** in **Presets** selection
>  
> - Select the **Platforms** tab and unselect **MacCatalyst** and **Windows** platforms
>  
> For more information on all the template options, see [Using the Uno Platform Template](xref:Uno.GettingStarted.UsingWizard).
>  
> - Click **Create** to complete the wizard
>  
> The template will create a solution with a single cross-platform project, named `DevExpressApp`, ready to run.
> 

Reviewer: ```suggestion
- Launch **Visual Studio** and click on **Create new project** on the Start Window. Alternatively, if you're already in Visual Studio, click **New, Project** from the **File** menu.
 
- Type `Uno Platform` in the search box
 
- Click **Uno Platform App**, then **Next**
 
- Name the project `DevExpressApp` and click **Create**
 
At this point you'll enter the **Uno Platform Template Wizard**, giving you options to customize the generated application.
 
- Select **Blank** in **Presets** selection
 
- Select the **Platforms** tab and unselect **MacCatalyst** and **Windows** platforms
 
- Click **Create** to complete the wizard
 
The template will create a solution with a single cross-platform project, named `DevExpressApp`, ready to run.

For more information on all the template options, see [Using the Uno Platform Template](xref:Uno.GettingStarted.UsingWizard).
```

### [Command Line](#tab/cli)
 
> [!NOTE]
> If you don't have the Uno Platform dotnet new templates installed, follow [dotnet new templates for Uno Platform](xref:Uno.GetStarted.dotnet-new).
 
Create a new application using the `unoapp` template, enabling .NET MAUI Embedding. In this case, we're going to use the Blank template (`-preset blank`) and include .NET MAUI Embedding support (`-maui`).
 
    ```bash

>     dotnet new unoapp -preset blank -maui -platforms "android" -platforms "ios" -platforms "wasm" -platforms "desktop" -o DevExpressApp

Reviewer: Does WASM and Desktop support MauiEmbedding?

Reviewer 2: I have made the adjustment @eriklimakc

```
 
This will create a new folder called **DevExpressApp** containing the new application.

---

## Nuget Packages

Add a reference to `DevExpress.Maui.DataGrid` to the DevExpressApp.MauiControls project.

## Updating AppBuilderExtensions

In the `AppBuilderExtensions` class, on `DevExpressApp.MauiControls` project, update the `UseMauiControls` extension method to call the `UseDevExpress` method.

>     ```cs

Reviewer: ```suggestion
    ```csharp
```

- Select **Blank** in **Presets** selection

- Select the **Platforms** tab and unselect **WebAssembly**, **macOS (Catalyst)**, **Windows**, and **Desktop** platforms

- Select the **Features** tab and click on **.NET MAUI Embedding** and **Toolkit**

- Click **Create** to complete the wizard

The template will create a solution with a single cross-platform project, named `DevExpressApp`, ready to run.

For more information on all the template options, see [Using the Uno Platform Template](xref:Uno.GettingStarted.UsingWizard).

### [Command Line](#tab/cli)

> [!NOTE]
> If you don't have the **Uno Platform dotnet new templates** installed, follow [dotnet new templates for Uno Platform](xref:Uno.GetStarted.dotnet-new).

Create a new application using the `unoapp` template, enabling .NET MAUI Embedding. In this case, we're going to use the Blank template (`-preset blank`) and include .NET MAUI Embedding support (`-maui`).

```bash
dotnet new unoapp -preset blank -maui -toolkit -platforms "android" -platforms "ios" -o DevExpressApp
```

This will create a new folder called **DevExpressApp** containing the new application.

---

> ## Nuget Packages

Reviewer 3: ```suggestion
## NuGet Packages
```

                .UseDevExpress()
                .ConfigureFonts(fonts =>
                {
                    fonts.AddFont("Assets/Fonts/OpenSansRegular.ttf", "OpenSansRegular");
                    fonts.AddFont("Assets/Fonts/OpenSansSemibold.ttf", "OpenSansSemibold");
                });
    }

>     ```

Reviewer: I think the snippet there isn't showing as it should:

![image](https://github.com/unoplatform/uno.extensions/assets/114886335/dd71d8f2-4318-4ab0-9950-5c61096fa70e)

    				<dxc:LineSeries.Data>
    					<dxc:SeriesDataAdapter DataSource="{Binding GdpValueForUSA.Values}"
    										   ArgumentDataMember="Year">
    						<dxc:ValueDataMember Type="Value"
    											 Member="Value" />
    					</dxc:SeriesDataAdapter>
    				</dxc:LineSeries.Data>
    			</dxc:LineSeries>

    			<dxc:LineSeries DisplayName="{Binding GdpValueForChina.CountryName}">
    				<dxc:LineSeries.Data>
    					<dxc:SeriesDataAdapter DataSource="{Binding GdpValueForChina.Values}"
    										   ArgumentDataMember="Year">
    						<dxc:ValueDataMember Type="Value"
    											 Member="Value" />
    					</dxc:SeriesDataAdapter>
    				</dxc:LineSeries.Data>
    			</dxc:LineSeries>

    			<dxc:LineSeries DisplayName="{Binding GdpValueForJapan.CountryName}">
    				<dxc:LineSeries.Data>
    					<dxc:SeriesDataAdapter DataSource="{Binding GdpValueForJapan.Values}"
    										   ArgumentDataMember="Year">
    						<dxc:ValueDataMember Type="Value"
    											 Member="Value" />
    					</dxc:SeriesDataAdapter>
    				</dxc:LineSeries.Data>
    			</dxc:LineSeries>
    		</dxc:ChartView.Series>
    	</dxc:ChartView>

> </ScrollView>
>     ```

Reviewer: ```suggestion
    </ScrollView>
    ```
```

>     ```xml
>     <?xml version="1.0" encoding="utf-8" ?>
>     <ScrollView xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
>                 xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
>                 xmlns:toolkit="http://schemas.microsoft.com/dotnet/2022/maui/toolkit"
>                 xmlns:dxg="clr-namespace:DevExpress.Maui.DataGrid;assembly=DevExpress.Maui.DataGrid"
>                 xmlns:dxc="clr-namespace:DevExpress.Maui.Charts;assembly=DevExpress.Maui.Charts"
>                 x:Class="DevExpressApp.MauiControls.ChartControl"
>                 xmlns:local="clr-namespace:DevExpressApp.MauiControls"
>                 HorizontalOptions="Fill"
>                 VerticalOptions="Fill">
>       <dxc:ChartView>
>     		<dxc:ChartView.BindingContext>
>     			<local:ViewModel />
>     		</dxc:ChartView.BindingContext>
>     		<dxc:ChartView.Series>
>     			<dxc:LineSeries DisplayName="{Binding GdpValueForUSA.CountryName}">
>     				<dxc:LineSeries.Data>
>     					<dxc:SeriesDataAdapter DataSource="{Binding GdpValueForUSA.Values}"
>     										   ArgumentDataMember="Year">
>     						<dxc:ValueDataMember Type="Value"
>     											 Member="Value" />
>     					</dxc:SeriesDataAdapter>
>     				</dxc:LineSeries.Data>
>     			</dxc:LineSeries>
> 
>     			<dxc:LineSeries DisplayName="{Binding GdpValueForChina.CountryName}">
>     				<dxc:LineSeries.Data>
>     					<dxc:SeriesDataAdapter DataSource="{Binding GdpValueForChina.Values}"
>     										   ArgumentDataMember="Year">
>     						<dxc:ValueDataMember Type="Value"
>     											 Member="Value" />
>     					</dxc:SeriesDataAdapter>
>     				</dxc:LineSeries.Data>
>     			</dxc:LineSeries>
> 
>     			<dxc:LineSeries DisplayName="{Binding GdpValueForJapan.CountryName}">
>     				<dxc:LineSeries.Data>
>     					<dxc:SeriesDataAdapter DataSource="{Binding GdpValueForJapan.Values}"
>     										   ArgumentDataMember="Year">
>     						<dxc:ValueDataMember Type="Value"
>     											 Member="Value" />
>     					</dxc:SeriesDataAdapter>
>     				</dxc:LineSeries.Data>
>     			</dxc:LineSeries>
>     		</dxc:ChartView.Series>
>     	</dxc:ChartView>
>     </ScrollView>
>     ```

Reviewer: ```suggestion
    ```xml
    <?xml version="1.0" encoding="utf-8" ?>
    <ScrollView xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
                xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
                xmlns:toolkit="http://schemas.microsoft.com/dotnet/2022/maui/toolkit"
                xmlns:dxg="clr-namespace:DevExpress.Maui.DataGrid;assembly=DevExpress.Maui.DataGrid"
                xmlns:dxc="clr-namespace:DevExpress.Maui.Charts;assembly=DevExpress.Maui.Charts"
                x:Class="DevExpressApp.MauiControls.ChartControl"
                xmlns:local="clr-namespace:DevExpressApp.MauiControls"
                HorizontalOptions="Fill"
                VerticalOptions="Fill">
        <dxc:ChartView>
            <dxc:ChartView.BindingContext>
                <local:ViewModel />
            </dxc:ChartView.BindingContext>
            <dxc:ChartView.Series>
                <dxc:LineSeries DisplayName="{Binding GdpValueForUSA.CountryName}">
                    <dxc:LineSeries.Data>
                        <dxc:SeriesDataAdapter DataSource="{Binding GdpValueForUSA.Values}"
                                               ArgumentDataMember="Year">
                            <dxc:ValueDataMember Type="Value"
                                                 Member="Value" />
                        </dxc:SeriesDataAdapter>
                    </dxc:LineSeries.Data>
                </dxc:LineSeries>

                <dxc:LineSeries DisplayName="{Binding GdpValueForChina.CountryName}">
                    <dxc:LineSeries.Data>
                        <dxc:SeriesDataAdapter DataSource="{Binding GdpValueForChina.Values}"
                                               ArgumentDataMember="Year">
                            <dxc:ValueDataMember Type="Value"
                                                 Member="Value" />
                        </dxc:SeriesDataAdapter>
                    </dxc:LineSeries.Data>
                </dxc:LineSeries>

                <dxc:LineSeries DisplayName="{Binding GdpValueForJapan.CountryName}">
                    <dxc:LineSeries.Data>
                        <dxc:SeriesDataAdapter DataSource="{Binding GdpValueForJapan.Values}"
                                               ArgumentDataMember="Year">
                            <dxc:ValueDataMember Type="Value"
                                                 Member="Value" />
                        </dxc:SeriesDataAdapter>
                    </dxc:LineSeries.Data>
                </dxc:LineSeries>
            </dxc:ChartView.Series>
        </dxc:ChartView>
    </ScrollView>
    ```
```

## Adding DataGridView Control

> 1. Update the `EmbeddedControl.xaml` in the  `DevExpressApp.MauiControls` project with the following XAML that includes the `DataGridView` control.

Reviewer: ```suggestion
1. Update the `EmbeddedControl.xaml` in the `DevExpressApp.MauiControls` project with the following XAML that includes the `DataGridView` control.
```

    ```

    > [!NOTE]

>     > You may notice that the `Binding` markup extension is used on some properties. The `MauiEmbedding` can handle bindings between Maui Controls and Uno Platform, just make sure the property in the `Binding` expression matches the property on your ViewModel.

Reviewer 3: ```suggestion
    > You may notice that the `Binding` markup extension is used on some properties. The `MauiEmbedding` can handle bindings between MAUI Controls and Uno Platform, just make sure the property in the `Binding` expression matches the property on your ViewModel.
```

    }
    ```

> 1. It's time to create the ViewModel that will hold the properties that will be data bound to the `DataGridViewControl` control. In `DevExpressApp` project, create a new folder called `ViewModels` and add a new class called `MainViewModel`. This class will have the following code:

Reviewer 3: ```suggestion
1. It's time to create the ViewModel that will hold the properties that will be data bound to the `DataGridViewControl` control. In the `DevExpressApp` project, create a new folder called `ViewModels` and add a new class called `MainViewModel`. This class will have the following code:
```

    ```

    > [!NOTE]
    > You may notice that the `Binding` markup extension is used on some properties. The `MauiEmbedding` can handle bindings between Maui Controls and Uno Platform, just make sure the property in the `Binding` expression matches the property on your ViewModel.

1. Update the `EmbeddedControl.xaml.cs` with the following code:

>     ```cs

Reviewer: ```suggestion
    ```csharp
```

    > [!NOTE]

>     > You may notice that the `Binding` markup extension is used on some properties. The `MauiEmbedding` can handle bindings between Maui Controls and UnoPlatform, just make sure the property in the `Binding` expression matches the property on your ViewModel.

Reviewer: ```suggestion
    > You may notice that the `Binding` markup extension is used on some properties. The `MauiEmbedding` can handle bindings between Maui Controls and Uno Platform, just make sure the property in the `Binding` expression matches the property on your ViewModel.
```

> 1. Update the `EmbeddedControl.xaml.cs` with the following code.

Reviewer: ```suggestion
1. Update the `EmbeddedControl.xaml.cs` with the following code:
```

    }
    ```

1. It's time to create the ViewModel that will hold the properties that will be data bound to the `DataGridViewControl` control. In `DevExpressApp` project, create a new folder called `ViewModels` and add a new class called `MainViewModel`. This class will have the following code:

>     ```cs

Reviewer: ```suggestion
    ```csharp
```
